### PR TITLE
Years fix cas in #154

### PR DESCRIPTION
**Pull request by layik**
Can add docs updates in this PR too.

### CODE REVIEW

**Comment by Robinlovelace** [comment]
Reprex showing that the class of the result with many years is different: 

``` r
devtools::install_github("ropensci/stats19", ref = "years")
#> Skipping install of 'stats19' from a github remote, the SHA1 (c9e46e67) has not changed since last install.
#>   Use `force = TRUE` to force installation
library(stats19)
#> Data provided under OGL v3.0. Cite the source and link to:
#> www.nationalarchives.gov.uk/doc/open-government-licence/version/3/
casualties_2013 = get_stats19(2013, type = "cas", silent = TRUE)
class(casualties_2013)
#> [1] "spec_tbl_df" "tbl_df"      "tbl"         "data.frame"
ncol(casualties_2013)
#> [1] 14
casualties_2014 = get_stats19(2014, type = "cas", silent = TRUE)
ncol(casualties_2014)
#> [1] 15
casualties_2013_2014 = get_stats19(2013:2014, type = "cas")
#> Files identified: DfTRoadSafety_Casualties_2013.zip
#>    http://data.dft.gov.uk.s3.amazonaws.com/road-accidents-safety-data/DfTRoadSafety_Casualties_2013.zip
#> Data already exists in data_dir, not downloading
#> Data saved at ~/stats19-data/DfTRoadSafety_Casualties_2013/DfTRoadSafety_Casualties_2013.csv
#> Files identified: DfTRoadSafety_Casualties_2014.zip
#>    http://data.dft.gov.uk.s3.amazonaws.com/road-accidents-safety-data/DfTRoadSafety_Casualties_2014.zip
#> Data already exists in data_dir, not downloading
#> Data saved at ~/stats19-data/DfTRoadSafety_Casualties_2014/DfTRoadSafety_Casualties_2014.csv
class(casualties_2013_2014)
#> [1] "data.frame"
```

<sup>Created on 2020-02-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by Robinlovelace** [comment]
Update, just tried on more years, good to see that this works: 

```
casualties_2013_2016 = get_stats19(2013:2016, type = "cas")
```

**Comment by layik** [comment]
Ok so however tibble is made needs to replace where I created "data.frame". Again can we avoid tibble? :) Or too late to ask?

**Comment by Robinlovelace** [comment]
> Again can we avoid tibble? :) Or too late to ask?

Never too late to ask but in this context the tibble print method has clear advantages, we already use tibble, so would break existing code if we purged tibble now. Replacing `data.frame` with `tibble::tibble` in your latest commit should solve the issue that class on the `years` branch changes depending on the number of years which is not ideal.

**Comment by Robinlovelace** [comment]
Update, I just noticed that we do not currently import tibble. But it's already there in readr so adding it as an Import will not change the size of the package:

https://github.com/ropensci/stats19/blob/master/DESCRIPTION

**Comment by layik** [comment]
Indeed and thanks for documenting it here for future us/others. 

Easy fix just can't get on it now.

**Comment by layik** [comment]
Right, lets be clear on few things:

1. `is.tibble` will [only](https://github.com/tidyverse/tibble/blob/master/R/tibble.R#L176) check this:
```r
is_tibble <- function(x) {
  inherits(x, "tbl_df")
}
```

2. [`as_tibble`](https://github.com/tidyverse/tibble/blob/e3eac8f83fc04ca31a5b007488b30bcb488d4cb8/R/as_tibble.R#L13) is `as.data.frame` to `data.frame` 

and it was previously tibble in our case so I think we can safely just add back the class without specifically changing dependencies. 

I am going to go ahead and do that, if we find a better solution or no option but to import `readr` then ... we will do so.

**Comment by layik** [comment]
The solution needs much more work. `stats19` is becoming almost too complicated for what it was in my mind.

Please hold this PR.

**Comment by layik** [comment]
that is an "explicit" way of replacing the `data.fram` statement. Please feel free to break it.

File: R/get.R [hunk 1]
```diff
@@ -146,7 +146,24 @@ get_stats19 = function(year = NULL,
...
+            diff.all = setdiff(names(all.keep.cols), names(y))
+            all.keep.cols[diff.all] = NA
+            # reciprocate
+            y[diff.all] = NA
+            all.keep.cols[diff.y] = NA
+            # now do.call rbind
+            all.keep.cols = do.call(rbind, list(y, all.keep.cols))
+        }
+      }
+      all = all.keep.cols
```
Comment (agila5) [inline R/get.R]: In my opinion the easiest solution is the following:

``` r
if (output_format == "ppp") {
      all = do.call(spatstat::superimpose, all)
} else {
     all_colnames = unique(unlist(lapply(all, names)))
     all = lapply(all, function(x) {
       x[setdiff(all_colnames, names(x))] <- NA
       x
     })
     all = do.call(rbind, all)
}
```

I think it should pass all the tests that you added.

Comment (layik) [reply R/get.R]: I am not sure which is "easiest" but I don't mind either. I would be happy with either. 

Less loops for me is best.

**Comment by Robinlovelace** [comment]
I think the suggestion by @agila5 looks good for spatstat objects. But would output of class `spatstat` ever have variable columns? I thought it was just the casualties .csv files that had variable column numbers depending on year but am not sure.

**Comment by agila5** [comment]
> I think the suggestion by @agila5 looks good for spatstat objects. But would output of class spatstat ever have variable columns? I thought it was just the casualties .csv files that had variable column numbers depending on year but am not sure.

I'm sorry but I'm not following you here because I didn't modify the if statement related to spatstat objects 🤔

**Comment by Robinlovelace** [comment]
Ah OK the `spatstat` bit was just for context. In that case, yes, this looks like a simple and good solution, but I've not tested it:

```r
all = lapply(all, function(x) {
       x[setdiff(all_colnames, names(x))] <- NA
       x
     })
```

**Comment by layik** [comment]
@agila5 please replace the chunk with your feed. I can do it of course.

**Comment by Robinlovelace** [comment]
Good to merge I think. Plan to do so tomorrow. Heads-up @agila5 and @layik, many thanks for this fix that turned out to be bigger than expected from the initial small issue. Mission has crept, but not far!

**Comment by layik** [comment]
so long as we keep dependencies to minimum, it is time well spent.

**Comment by Robinlovelace** [comment]
I breathed a sigh of relief when the CI build passed (partly thanks to few deps as recommended wisely by @layik ). But there are still issues due to upstream external dependencies: #158

**Pull request merged**